### PR TITLE
[java] Merge rules that check for empty things

### LINKED ISSUE #3096: [java] EmptyStatementNotInLoop FP in 6.30.0 with IfStatement
**Affects PMD Version:** 6.30.0

**Rule:** EmptyStatementNotInLoop
https://pmd.github.io/pmd-6.30.0/pmd_rules_java_errorprone.html#emptystatementnotinloop

**Description:** When there is Empty If Statement, it gives alarms.
However, there is another rule which could catch Empty If Statement.
I think it would generate duplicated alarms 
with the same code and the same reason.

**Code Sample demonstrating the issue:**

```
class TestEmptyStatementNotInLoop{
        public static void main(String[] args){
                String s = "test";
                if(s == null);
        }
}
```

**Expected outcome:**
Don't give any alarm.

**Running PMD through:** *[CLI]*

### PR DESCRIPTION
**Pull request by oowekyala**
### New rules

- EmptyControlStatement: Reports many instances of empty things, e.g. control statements whose 
  body is empty, as well as empty initializers.
  
  EmptyControlStatement also works for empty `for` and `do` loops, while there were previously 
  no corresponding rules.

  Replaces rules EmptyFinallyBlock, EmptyIfStmt, EmptyInitializer, EmptyStatementBlock,
  EmptySwitchStatements, EmptySynchronizedBlock, EmptyTryBlock, and EmptyWhileStmt.

- UnnecessarySemicolon: Reports semicolons that are unnecessary  (so called "empty statements" 
  and "empty declarations").
  
  Replaces the rule EmptyStatementNotInLoop.

Both new rules are in category codestyle, while the replaced rules are in category errorprone. 

Note that *EmptyCatchBlock* is still an independent rule (maybe it shouldn't? ~I'm leaning towards merging it into the rule now, let me know what you think~ I think EmptyCatchBlock should generally have higher priority than EmptyControlStatement so it's probably best to keep it separate). :+1:

### Deprecated rules

* EmptyFinallyBlock
* EmptyIfStmt
* EmptyInitializer
* EmptyStatementBlock
* EmptySwitchStatements
* EmptySynchronizedBlock
* EmptyTryBlock
* EmptyWhileStmt
* EmptyStatementNotInLoop

## Related issues

- Fixes #3096

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)
- [x] Deprecate old rules
- [x] New rules should be part of quickstart ruleset
- [x] Release notes

### CODE REVIEW

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 1]
```diff
@@ -710,6 +710,50 @@ public abstract class ShouldBeAbstract {
         </example>
     </rule>
 
+    <rule name="EmptyControlStatement"
+          language="java"
+          since="6.42.0"
+          message="This control statement has an empty branch"
+          class="net.sourceforge.pmd.lang.java.rule.codestyle.EmptyControlStatementRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_codestyle.html#unnecessarysemicolon">
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml:718]:
```suggestion
          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_codestyle.html#emptycontrolstatement">
```

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 2]
```diff
@@ -710,6 +710,50 @@ public abstract class ShouldBeAbstract {
...
+            Reports control statements whose body is empty, as well as empty initializers.
+
+            The checked code constructs are the following:
+            - bodies of `try` statements (not try-with-resources)
+            - `finally` clauses of `try` statements
+            - `switch` statements
+            - `synchronized` statements
+            - `if` statements
+            - loop statements: `while`, `for`, `do .. while`
+            - initializers
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml:733]:
```suggestion
            - initializers
            - blocks used as statements (for scoping)
```

**Review by adangel — COMMENTED** [review]
I think, we should simply merge it and see, what feedback we get. It doesn't break things - the old rules will just be deprecated. And in worst case, we can undeprecate the rules again.

I'd also don't add properties to enable/disable checks. That would be one advantage of having several rules - you can enable/disable single checks. But let's wait if anyone actually needs this configurability.

I'll update this branch and finish it.

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 3]
```diff
@@ -710,6 +710,50 @@ public abstract class ShouldBeAbstract {
...
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_codestyle.html#emptycontrolstatement">
+        <!-- Note about naming: EmptyCodeBlock does not work because the rule flags `if(true);`  -->
+        <!-- EmptyControlStatement is weird because `{}` is not a control statement, and we also flag initializers. -->
+        <!-- EmptyStatement does not work because `;` is called an "empty statement" and is not flagged by this rule, rather, by UnnecessarySemicolon. -->
+        <!-- EmptyCodeConstruct would work but sounds a bit odd (right?). -->
+        <description><![CDATA[
+            Reports control statements whose body is empty, as well as empty initializers.
+
+            The checked code constructs are the following:
+            - bodies of `try` statements (not try-with-resources)
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/codestyle.xml:727]:
```suggestion
            - bodies of `try` statements
```

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 3291 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/01d94b00a654455192b199079e90f3d2/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**